### PR TITLE
Fix Catalogs plugin for recent changes to astroquery

### PR DESCRIPTION
**Pull request by ejeschke**
It seems that the API for astroquery VO conesearch changed.  This fixes up the `Catalogs` plugin so that it works again.

I also removed the `pyvo` interface. Not sure if this package is being deprecated in favor of `astroquery` for VO cone searches. It can be added back if necessary.

Fix: #52

### CODE REVIEW

**Comment by ejeschke** [comment]
No hurry, @pllim.  Curious though whether the VO conesearch provides standard columns for object name, ra and dec.

**Comment by pllim** [comment]
Actually... The import should be updated too to this:
```python
# Do we have astroquery >=0.3.5 installed?
from astropy import coordinates, units

have_astroquery = False
try:
    from astroquery.vo_conesearch import conesearch
    have_astroquery = True
except ImportError:
    pass
```

The old astropy import is no longer applicable.

**Comment by ejeschke** [comment]
> As for column names, that depends on the individual cone search service providers and is out of control of the astroquery.vo_conesearch interface. To have a mapping of all possible ways they name their RA and Dec columns would be unmaintainable.

So, if you leave out the catalog parameter, it's a toss-up as to what the column names are coming back?

**Comment by pllim** [comment]
Yes, depends on the service you are calling (`full_name`) and even if you only use that one service, there is no guarantee that they won't change things up on server side.

**Comment by ejeschke** [comment]
I thought that one of the purposes of VO was to standardize things. I guess you have standardized the conesearch query side in `astroquery` (the query), but not the results coming back.

**Comment by pllim** [comment]
> I thought that one of the purposes of VO was to standardize things

🤐

**Comment by pllim** [comment]
https://xkcd.com/927/

**Comment by ejeschke** [comment]
In that case, @pllim, I think we should try to cover the most popular services returned by `conesearch.list_catalogs()`. Basically, we need to know the id, ra and dec returned in the table--all fields are added anyway tto the table and the magnitude field can be selected by the user. It would be desirable to auto-detect some of the more common magnitude fields, but not necessary.

How familiar are you with the id, ra, and dec fields of the dozen or so services returned by `list_catalogs()`?

**Comment by pllim** [comment]
Well... If you are thinking of https://astroquery.readthedocs.io/en/latest/vo_conesearch/vo_conesearch.html#default-cone-search-services , it is really outdated and I don't have time or motivation to update them (ref astropy/astroquery#871 ).

What would be more practical is for you to identify services that are useful for Ginga (i.e., what do you use it for and what catalogs do you really use?) and go from there.

**Comment by ejeschke** [comment]
> What would be more practical is for you to identify services that are useful for Ginga (i.e., what do you use it for and what catalogs do you really use?) and go from there.

From Subaru we use our own catalog service (i.e. not `AstroPyCatalogServer`) but that has a matching API (and we also subclass `Catalogs`).  For generic reference viewer, I'm thinking to maybe provide a useful set of catalogs by default that are available via `astroquery`.  Maybe I should just select a half dozen that are broadly useful and add those names to the default configuration in this PR.

**Comment by ejeschke** [comment]
@pllim, please check out recent changes. We are seeing some of the fruits of recent PRs! With #862, we can now define catalog sources in the `plugins_Catalogs.cfg` file.  I set some default sources, see if you agree with the selection.

File: ginga/examples/configs/plugin_Catalogs.cfg [hunk 1]
```diff
@@ -7,3 +7,29 @@ draw_type = 'circle'
...
 color_outline = 'aquamarine'
 click_radius = 10
+
+# catalog sources
+# Format: list of dicts
+# Each dict defines a source, and has the following fields:
+#   shortname: str, the short name appearing in the control for selecting a source
+#               in the plugin
+#   fullname: str, the full name, should correspond *exactly* with the name required
+#               by astroquery.vo.conesearch "catalog" parameter
```
Comment (pllim) [inline ginga/examples/configs/plugin_Catalogs.cfg:17]:
```suggestion
#               by astroquery.vo_conesearch "catalog" parameter
```

File: ginga/examples/configs/plugin_Catalogs.cfg [hunk 2]
```diff
@@ -7,3 +7,29 @@ draw_type = 'circle'
...
+#   type: str, should be "astroquery.vo" for an astroquery.vo.conesearch function
```
Comment (pllim) [inline ginga/examples/configs/plugin_Catalogs.cfg:18]:
```suggestion
#   type: str, should be "astroquery.vo_conesearch" for an astroquery.vo_conesearch function
```

File: ginga/examples/configs/plugin_Catalogs.cfg [hunk 3]
```diff
@@ -7,3 +7,29 @@ draw_type = 'circle'
...
+#   mapping: a nested dict providing the mapping for the return results to the GUI,
+#            in terms of field name to Ginga table.
+#            There must be keys for 'id', 'ra' and 'dec'. 'mag', if present, can be
+#            a list of field names that define magnitudes of the elements in various
+#            wavelengths.
+#
+catalog_sources = [
+    {'shortname': "GSC 2.3", 'fullname': "Guide Star Catalog 2.3 Cone Search 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/examples/configs/plugin_Catalogs.cfg:27]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/examples/configs/plugin_Catalogs.cfg [hunk 4]
```diff
@@ -7,3 +7,29 @@ draw_type = 'circle'
...
+     'mapping': {'id': 'objID', 'ra': 'ra', 'dec': 'dec', 'mag': ['Mag']}},
+    {'shortname': "USNO-A2.0 1", 'fullname': "The USNO-A2.0 Catalogue (Monet+ 1998) 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/examples/configs/plugin_Catalogs.cfg:30]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/examples/configs/plugin_Catalogs.cfg [hunk 5]
```diff
@@ -7,3 +7,29 @@ draw_type = 'circle'
...
+     'mapping': {'id': 'USNO-A2.0', 'ra': 'RAJ2000', 'dec': 'DEJ2000', 'mag': ['Bmag', 'Rmag']}},
+    {'shortname': "2MASS 1", 'fullname': "Two Micron All Sky Survey (2MASS) 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/examples/configs/plugin_Catalogs.cfg:33]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/rv/plugins/Catalogs.py [hunk 6]
```diff
@@ -9,40 +9,33 @@
...
-channel you want to perform searchs on. You will see the catalogs listed
-in a drop-down menu on the plugin GUI.
-
 Draw a shape on the displayed image and adjust search parameters as desired.
 When you are ready, press on the button to perform the search.
 When search results are available, they will be displayed on the image and
 also listed in a table on the plugin GUI. You can click on either the table
 or the image to highlight selection.
 
+**Defining catalogs**
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:18]:
```suggestion
**Defining catalogs**

```

File: ginga/rv/plugins/Catalogs.py [hunk 7]
```diff
@@ -9,40 +9,33 @@
...
+You can customize the catalogs used by copying the `plugin_Catalogs.cfg`
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:19]:
```suggestion
You can customize the catalogs used by copying the ``plugin_Catalogs.cfg``
```

File: ginga/rv/plugins/Catalogs.py [hunk 8]
```diff
@@ -9,40 +9,33 @@
...
+from the example configurations directory to your $HOME/.ginga directory and
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:20]:
```suggestion
from the example configurations directory to your ``$HOME/.ginga`` directory and
```

File: ginga/rv/plugins/Catalogs.py [hunk 9]
```diff
@@ -9,40 +9,33 @@
...
+modifying the `catalog_sources` table defined inside.
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:21]:
```suggestion
modifying the ``catalog_sources`` table defined inside.
```

File: ginga/rv/plugins/Catalogs.py [hunk 10]
```diff
@@ -9,40 +9,33 @@
...
+
+The format of this option is a list of dicts, where each dict defines a
+source and has the following fields:
+
+* shortname: str, the short name appearing in the control for selecting
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:26]:
```suggestion
* ``shortname``: str, the short name appearing in the control for selecting
```

File: ginga/rv/plugins/Catalogs.py [hunk 11]
```diff
@@ -9,40 +9,33 @@
...
+  a source in the plugin.  This is the name that will appear in the
+  drop-down menu in the catalogs control area of the GUI
+* fullname: str, the full name, should correspond *exactly* with the
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:29]:
```suggestion
* ``fullname``: str, the full name, should correspond *exactly* with the
```

File: ginga/rv/plugins/Catalogs.py [hunk 12]
```diff
@@ -9,40 +9,33 @@
...
+  name required by the `~astroquery.vo.conesearch` `catalog` parameter
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:30]: We don't have intersphinx set up for `astroquery`.
```suggestion
  name required by the ``astroquery.vo_conesearch`` ``catalog`` parameter
```

File: ginga/rv/plugins/Catalogs.py [hunk 13]
```diff
@@ -9,40 +9,33 @@
...
+* type: str, should be "astroquery.vo" for an astroquery.vo.conesearch
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:31]:
```suggestion
* ``type``: str, should be "astroquery.vo_conesearch" for an ``astroquery.vo_conesearch``
```

File: ginga/rv/plugins/Catalogs.py [hunk 14]
```diff
@@ -9,40 +9,33 @@
...
+  function
+* mapping: dict, a nested dict providing the mapping for the return results
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:33]:
```suggestion
* ``mapping``: dict, a nested dict providing the mapping for the return results
```

File: ginga/rv/plugins/Catalogs.py [hunk 15]
```diff
@@ -9,40 +9,33 @@
...
+  to the GUI, in terms of field name to Ginga table.  In each source dict,
+  there must be keys for 'id', 'ra' and 'dec'. 'mag', if present, can be a
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:35]:
```suggestion
  there must be keys for 'id', 'ra', and 'dec'. 'mag', if present, can be a
```

File: ginga/rv/plugins/Catalogs.py [hunk 16]
```diff
@@ -53,12 +46,28 @@ def post_gui_config(ginga):
...
-from ginga.util import wcs
+from ginga.util import wcs, catalog
 from ginga.gw import ColorBar, Widgets
 
 __all__ = ['Catalogs']
 
 
+default_catalog_sources = [
+    {'shortname': "GSC 2.3", 'fullname': "Guide Star Catalog 2.3 Cone Search 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:57]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/rv/plugins/Catalogs.py [hunk 17]
```diff
@@ -53,12 +46,28 @@ def post_gui_config(ginga):
...
+     'mapping': {'id': 'objID', 'ra': 'ra', 'dec': 'dec', 'mag': ['Mag']}},
+    {'shortname': "USNO-A2.0 1", 'fullname': "The USNO-A2.0 Catalogue (Monet+ 1998) 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:60]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/rv/plugins/Catalogs.py [hunk 18]
```diff
@@ -53,12 +46,28 @@ def post_gui_config(ginga):
...
+     'mapping': {'id': 'USNO-A2.0', 'ra': 'RAJ2000', 'dec': 'DEJ2000', 'mag': ['Bmag', 'Rmag']}},
+    {'shortname': "2MASS 1", 'fullname': "Two Micron All Sky Survey (2MASS) 1",
+     'type': 'astroquery.vo',
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:63]:
```suggestion
     'type': 'astroquery.vo_conesearch',
```

File: ginga/rv/plugins/Catalogs.py [hunk 19]
```diff
@@ -112,6 +123,38 @@ def __init__(self, fv, fitsimage):
...
         self.gui_up = False
 
+        bank = self.fv.get_ServerBank()
+
+        # add catalogs found in configuration file
+        catalogs = self.settings.get('catalog_sources', [])
+        for d in catalogs:
+            typ = d.get('type', None)
+            obj = None
+            if typ == 'astroquery.vo':
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:133]:
```suggestion
            if typ == 'astroquery.vo_conesearch':
```

File: ginga/rv/plugins/Catalogs.py [hunk 20]
```diff
@@ -112,6 +123,38 @@ def __init__(self, fv, fitsimage):
...
+                obj = catalog.AstroQueryVOCatalogServer(self.logger, d['fullname'],
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:134]: It is offically "Astroquery" and not "AstroQuery" according to their documentation.
```suggestion
                obj = catalog.AstroqueryVOCatalogServer(self.logger, d['fullname'],
```

File: ginga/util/catalog.py [hunk 21]
```diff
@@ -63,21 +46,19 @@ def has_key(self, key):
         return key in self.starInfo
 
 
-# TODO: Deprecate this class name and replace with something that reflects
-#       Cone Search in Astroquery.
-class AstroPyCatalogServer(object):
+class AstroQueryVOCatalogServer(object):
```
Comment (pllim) [inline ginga/util/catalog.py:49]: This renaming is certainly not backward-compatible?
```suggestion
class AstroqueryVOCatalogServer(object):
```

File: ginga/rv/plugins/Catalogs.py [hunk 22]
```diff
@@ -112,6 +123,38 @@ def __init__(self, fv, fitsimage):
...
+            if obj is not None:
+                bank.addCatalogServer(obj)
+
+        # add image servers found in configuration file
+        catalogs = self.settings.get('image_sources', [])
+        for d in catalogs:
+            typ = d.get('type', None)
+            obj = None
+            if typ == 'astroquery':
+                obj = catalog.AstroQueryVOCatalogServer(self.logger, d['fullname'],
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:149]:
```suggestion
                obj = catalog.AstroqueryVOCatalogServer(self.logger, d['fullname'],
```

File: ginga/util/catalog.py [hunk 23]
```diff
@@ -63,21 +46,19 @@ def has_key(self, key):
...
+    """For queries using the `astroquery.vo.conesearch` function."""
```
Comment (pllim) [inline ginga/util/catalog.py:50]:
```suggestion
    """For queries using the ``astroquery.vo_conesearch`` function."""
```

File: ginga/util/catalog.py [hunk 24]
```diff
@@ -200,6 +179,7 @@ def search(self, **params):
 
 
 class AstroQueryImageServer(object):
```
Comment (pllim) [inline ginga/util/catalog.py:181]: Ah, didn't realize it is also "AstroQuery" and not "Astroquery" here but maybe this is out of scope.

File: ginga/util/catalog.py [hunk 25]
```diff
@@ -200,6 +179,7 @@ def search(self, **params):
...
+    """For queries using the `astroquery.vo.conesearch` function."""
```
Comment (pllim) [inline ginga/util/catalog.py:182]:
```suggestion
    """For queries using the ``astroquery.vo_conesearch`` function."""
```

File: ginga/rv/plugins/Catalogs.py (hunk 15 above)
Comment (ejeschke) [reply ginga/rv/plugins/Catalogs.py:35]: You and your oxford commas! :smile:

File: ginga/util/catalog.py (hunk 24 above)
Comment (ejeschke) [reply ginga/util/catalog.py:181]: We can provide some backward compatible classes that are just wrappers around the new class names and constructors, I think.

**Comment by ejeschke** [comment]
@pllim, when you have a chance, can you check that your `ginga_config.py` catalog additions still work with this?  I updated instructions for plugin (please give your feedback!), will add an updated screenshot.  Also, if you have any recommendations for default catalogs to add LMK.

Aside from screenshot, I think this is ready to merge.  We can always update default sources later.

**Comment by pllim** [comment]
I'll have a look soon. Thanks!

**Comment by pllim** [comment]
@ejeschke , looks like there is a conflict that needs resolving.

**Comment by ejeschke** [comment]
Should be fixed now. Let's see how the CI turns out...

**Comment by pllim** [comment]
Plugging in the old setup in `~/.ginga/ginga_config.py`:
```python
def pre_gui_config(ginga):
    from ginga.util.catalog import AstroPyCatalogServer

    # Add Cone Search services
    catalogs = [
        ('The HST Guide Star Catalog, Version 1.2 (Lasker+ 1996) 1',
         'GSC_1.2'),
        ('The PMM USNO-A1.0 Catalogue (Monet 1997) 1', 'USNO_A1'),
        ('The USNO-A2.0 Catalogue (Monet+ 1998) 1', 'USNO_A2'),
    ]
    bank = ginga.get_ServerBank()
    for longname, shortname in catalogs:
        obj = AstroPyCatalogServer(
            ginga.logger, longname, shortname, '', shortname)
        bank.addCatalogServer(obj)
```

I get this error when trying to "Search Catalog": `Query exception: Failed to load catalog: could not convert string to float: ''`

My comment from Aug 6 review still stands, "I think there is still the matter of change log and backward-compatibility." That is, the change log should concisely summarize this PR. Or if more explanation is needed, perhaps add a new section in the documentation that answers the following questions:

* Can I use the old `~/.ginga/ginga_config.py`?
* If not, is there a conversion guide to help me translate the old setup to the new setup?
* What have changed in this new API that I need to know about as an end-user? Are there new settings? Are there old settings that no longer works?

Thanks!

**Comment by pllim** [comment]
p.s. I wonder if the screenshots need to be so hi-res. Perhaps can think about downscaling them to something that is acceptable for the 800-pixel wide setting. If you choose to do that, you will also want to squash out the hi-res versions from commit history.

**Comment by ejeschke** [comment]
I will definitely update the notes in the `WhatsNew`: the primary purpose of the PR is to fix the calls to the `Astroquery` API, which has changed.   I am trying to make sure it _is_ backward compatible, so that it doesn't need any lengthy conversion guide. I will mention that we can now define the catalog entries in the `plugin_Catalogs.cfg` and to follow the example one (which has been updated). 

A secondary purpose is to finally provide some default values for catalog and image sources.

**Comment by ejeschke** [comment]
Rebased.

**Comment by ejeschke** [comment]
@pllim, as we are approaching the 1 year mark since the last Ginga release (:disappointed: ), I'm trying to get this PR wrapped up, merged and make a 3.2 release.

I've decided that I will not attempt to preserve backward compatible `Catalogs` configuration, including via `ginga_config.py`.  I think it is preferable to focus users on configuring via the `plugin_Catalogs.cfg` file.

I updated the `WhatsNew` *and* the plugin documentation itself to mention the non backward compatibility.
Take a look at the doc and let me know if it meets your criteria above.  At least users will have a working catalogs plugin (for those that use it) since it has been broken for quite a while, judging from the date this PR was first submitted!

I think this will make the Catalogs much more usable.  I suspect there will be more tweaks down the road, esp. changing the default catalog and image sources, perhaps.

**Comment by ejeschke** [comment]
If you think the screenshots are still too big, I can redo those to be a bit smaller.

File: ginga/rv/plugins/Catalogs.py [hunk 26]
```diff
@@ -8,41 +8,62 @@
 ``Catalogs`` is a local plugin, which means it is associated with a
 channel.  An instance can be opened for each channel.
 
+.. note:: to use ``Catalogs`` it is necessary to install the `astroquery`
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:11]: We don't have intersphinx to astroquery, so need double ticks.
```suggestion
.. note:: To use ``Catalogs``, it is necessary to install the ``astroquery``
```

File: ginga/rv/plugins/Catalogs.py [hunk 27]
```diff
@@ -8,41 +8,62 @@
...
+          package.
+
+.. note:: configuration of ``Catalogs`` via `ginga_config.py` technique
```
Comment (pllim) [inline ginga/rv/plugins/Catalogs.py:14]: I think this is important enough to be a warning.
```suggestion
.. warning:: Configuration of ``Catalogs`` via ``ginga_config.py`` technique in Ginga 3.2 or later
```

**Review by pllim — COMMENTED** [review]
`astroquery` needs to be added to optional dependency section in `setup.cfg`; `install.rst` also needs to be updated.

If breaking change is unavoidable, it is fine, though I want to test your release candidate downstream before the release actually happens.

cc @obi-wan76 in case TEL uses this `Catalogs` plugin.

**Comment by ejeschke** [comment]
> astroquery needs to be added to optional dependency section in setup.cfg

Do you mean the "options.extras_require" area (recommended=...)?  That looks like what is used for the other plugins.

**Comment by pllim** [comment]
> options.extras_require

Yes, that's it. 😸

**Comment by ejeschke** [comment]
@pllim, have a look at the `setup.cfg` and `install.rst` files; I believe I have incorporated the recommendations correctly.

**Review by pllim — APPROVED** [review]
LGTM FFTM. Thanks!

**Comment by ejeschke** [comment]
Thanks, @pllim. Let's coordinate the release in #867
